### PR TITLE
Fix `analyze-exclusions` crashes if there is no `dependencyManagement` element

### LINKED ISSUE #1474: [MDEP-968] analyze-exclusions crashes if there is no dependencyManagement element
**[Jakub Bochenski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbochenski)** opened **[MDEP-968](https://issues.apache.org/jira/browse/MDEP-968?redirect=false)** and commented

```
Execution analyze-exclusions of goal org.apache.maven.plugins:maven-dependency-plugin:3.8.1:analyze-exclusions failed: Cannot invoke "org.apache.maven.model.DependencyManagement.getDependencies()" because the return value of "org.apache.maven.project.MavenProject.getDependencyManagement()" is null
```

---
No further details from [MDEP-968](https://issues.apache.org/jira/browse/MDEP-968?redirect=false)

### PR DESCRIPTION
**Pull request by JackPGreen**
As reported in https://github.com/apache/maven-dependency-plugin/issues/1474, when executed on a project without a `dependencyManagement` tag defined, an exception is thrown:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.10.0:analyze-exclusions (default-cli) on project reproducer: Execution default-cli of goal org.apache.maven.plugins:maven-dependency-plugin:3.10.0:analyze-exclusions failed: Cannot invoke "org.apache.maven.model.DependencyManagement.getDependencies()" because the return value of "org.apache.maven.project.MavenProject.getDependencyManagement()" is null -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.10.0:analyze-exclusions (default-cli) on project reproducer: Execution default-cli of goal org.apache.maven.plugins:maven-dependency-plugin:3.10.0:analyze-exclusions failed: Cannot invoke "org.apache.maven.model.DependencyManagement.getDependencies()" because the return value of "org.apache.maven.project.MavenProject.getDependencyManagement()" is null
{...}
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.maven.model.DependencyManagement.getDependencies()" because the return value of "org.apache.maven.project.MavenProject.getDependencyManagement()" is null
    at org.apache.maven.plugins.dependency.exclusion.AnalyzeExclusionsMojo.execute (AnalyzeExclusionsMojo.java:106)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:919)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:285)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:207)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
```

When executing `analyze-exclusions`, we simply assume `getDependencyManagement` returns non-`null`. Whereas in other mojos a more cautious approach is taken, e.g.: https://github.com/apache/maven-dependency-plugin/blob/a99bc570747b2623a4f3e5ca5a3dc98e379bd58b/src/main/java/org/apache/maven/plugins/dependency/analyze/AnalyzeDepMgt.java#L118-L123

Updated this mojo (and corresponding test) to suit.

Fixes: https://github.com/apache/maven-dependency-plugin/issues/1474

- [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

### CODE REVIEW

File: src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java [hunk 1]
```diff
@@ -85,9 +83,7 @@ void setUp() {
         when(project.getArtifactId()).thenReturn("testArtifactId");
         when(project.getVersion()).thenReturn("1.0.0");
 
-        DependencyManagement dependencyManagement = mock(DependencyManagement.class);
-        when(dependencyManagement.getDependencies()).thenReturn(Collections.emptyList());
-        when(project.getDependencyManagement()).thenReturn(dependencyManagement);
+        when(project.getDependencyManagement()).thenReturn(null);
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java]: I think the existing tests should still work without changes. You need to add a new test that specifically addresses the null case.

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java (hunk 1 above)
Comment (JackPGreen) [reply src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java]: > I think the existing tests should still work without changes. You need to add a new test that specifically addresses the null case.

I did originally try to add a new explicit test case to cover this, but I realised the real issue is the mock test fixture is the real issue.

In the test we are mocking `getDependencyManagement()` to always return a non-`null` value - but the documentation offers no non-`null` guarantees and indeed some other internal Maven code is checking for `null` from this methods' invocation.

Comment (elharo) [reply src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java]: yes, but it doesn't say it's null either. Both cases, null and non-null, seem valid and need tests.

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java (hunk 1 above)
Comment (JackPGreen) [reply src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java]: > yes, but it doesn't say it's null either. Both cases, null and non-null, seem valid and need tests.

Added new test(s) in https://github.com/apache/maven-dependency-plugin/pull/1597/commits/9cb28fb8f511dd6a8956d66e719b0bbb64004e00

File: src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java [hunk 2]
```diff
@@ -197,6 +197,29 @@ void testThatLogContainProjectName(AnalyzeExclusionsMojo mojo) throws Exception
         verify(testLog).warn("projectName defines following unnecessary excludes");
     }
 
+    /**
+     * Nullability behavior of {@link MavenProject#getDependencyManagement} is not documented, test mojo with both {@code null}
+     * and non-{@code null} outputs
+     *
+     * @see <a href="https://github.com/apache/maven-dependency-plugin/issues/1474">Issue</a>
+     */
+    @ParameterizedTest
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java]: This feels overcomplicated for two cases and is really hard to follow. Can we just have two explicit test methods, one for each case, no parameterization?

Comment (JackPGreen) [reply src/test/java/org/apache/maven/plugins/dependency/exclusion/AnalyzeExclusionsMojoTest.java]: Sure - d0178300364ef152d790a4814cc9cb3d98f5c7bb

**Review by elharo — APPROVED** [review]

**Review by cstamas — APPROVED** [review]
